scanner.py

```diff
@@ -427,117 +510,591 @@ def start_touch(self, gcmd, touch_settings, verbose):
+
+                try:
+                    probe_position = self.phoming.probing_move(
+                        self.mcu_probe, homing_position, speed
+                    )
+                except self.printer.command_error as e:
+                    if self.printer.is_shutdown():
+                        self.trigger_method = 0
+                        raise self.printer.command_error(
+                            "Touch procedure interrupted due to printer shutdown"
+                        ) from e
+                    raise
+                finally:
+                    self.set_accel(max_accel)
+
+                retract_position = self.toolhead.get_position()[:]
+                retract_position[2] = min(retract_position[2] + retract_dist, z_max)
+                self.toolhead.move(retract_position, retract_speed)
+                self.toolhead.dwell(1.0)
+
+                samples.append(probe_position[2])
+                self.log_debug_info(
+                    verbose,
+                    gcmd,
+                    f"Touch {len(samples)} result: {probe_position[2]:.4f}",
+                )
+
+                average = np.median(samples)
+                deviation = max(abs(sample - average) for sample in samples)
+                deviation = round(deviation, 4)
+                if deviation > tolerance:
+                    self.log_debug_info(
+                        verbose,
+                        gcmd,
+                        f"Deviation of {deviation:.4f} exceeds tolerance of {tolerance:.4f}",
+                    )
+                    gcmd.respond_info(f"Attempt {retries + 1}/{max_retries} failed with deviation {deviation:.4f}.")
+                    retries += 1
+                    new_retry = True
+                    samples.clear()
+                    # If successful, we continue gathering samples until num_samples is reached.
+
+                self.log_debug_info(
+                    verbose,
+                    gcmd,
+                    f"Deviation: {deviation:.4f}\nNew Average: {average:.4f}\nTolerance: {tolerance:.4f}",
+                )
+
+            std_dev = np.std(samples) if samples else None
+            if len(samples) == num_samples:
+                success = True
+                position_difference = (
+                    initial_position[2] - self.toolhead.get_position()[2]
+                )
+                adjusted_difference = initial_position[2] - np.median(samples)
+                self.log_debug_info(
+                    verbose,
+                    gcmd,
+                    f"Position Difference: {position_difference:.4f}\nAdjusted Difference: {adjusted_difference:.4f}",
+                )
+            else:
+                std_dev = None
+                success = False
+
+            self.toolhead.wait_moves()
+            self.toolhead.flush_step_generation()
+            self.previous_probe_success = 1 if success else 0
+
+            # Return relevant data
```

Comment:
Should we set `self.detect_threshold_z` to `original_threshold` again? We could do it in a `finally` clause. The entire function is wrapped in a `try`.

Reply:
L861

Reply:
That is in a different function though?

Reply:
yes its in `cmd_SCANNER_THRESHOLD_SCAN` which loops `start_threshold_scan`

So once theyre finished it sets the `self.detect_threshold_z` back

Reply:
Alrighty! I'll accept that 👍